lib/actions/appleMusic.js

```diff
@@ -13,46 +16,177 @@ function getAlbumUri(id) {
+                return track.artistName.toLowerCase();
+              }).sort();
+              var tracks = tracksJson.results.map(function(track) {
+                return track.trackName.toLowerCase();
+              }).sort();
+
+              // Determine if the search results are many tracks for an artist OR a specific song title.
+              var prevArtist=artists[0];
+              var prevTrack=tracks[0];
+  
+              for (var i=1; i < tracksJson.resultCount;i++) {
+                if (artists[i] != prevArtist) {
+                  artistCount++;
+                  prevArtist = artists[i];
+                }
+                if (tracks[i] != prevTrack) {
+                  trackCount++;
+                  prevTrack = tracks[i];
+                }  
+              }
+          
+              // If there are at least twice as many track differences that artist differences, then load all of the artist's tracks into the queue, ELSE play just the first track now (iTunes returns the most popular track first) 
+              if (trackCount/artistCount > 2) {
+                const queueURI = 'x-rincon-queue:' + player.uuid + '#0';
+                
+                // Filtered list of tracks to play
+                var queueTracks = tracksJson.results.reduce(function(tracksArray, track) {
+                  if (track.isStreamable) {
+                    var skip = false;
+  
+                    for (var j=0; (j < tracksArray.length) && !skip ; j++) {
+                      // Skip duplicate songs
+                      skip = (track.trackName == tracksArray[j]);
+                    }
+                    
+                    if (!skip) {
+                      var metadataID = METADATA_URI_STARTERS[type] + encodeURIComponent('song:' + track.trackId);
+                      var metadata = getMetadata(metadataID, PARENTS[type], CLASSES[type],'');
+                      var uri = uriTemplates[type](encodeURIComponent('song:' + track.trackId));
+
+                      tracksArray.push({trackName:track.trackName, uri:uri, metadata:metadata});
+                    }
+                  } 
+                  return tracksArray;
+                }, []);
+
+                nextTrackNo = 0;
+                return player.coordinator.clearQueue()
+                  .then(() => player.coordinator.setAVTransport(queueURI, ''))
+                  .then(() => player.coordinator.addURIToQueue(queueTracks[0].uri, queueTracks[0].metadata, true, 1))
+                  .then(() => player.coordinator.play())
+                  .then(() => {
+                    return queueTracks.reduce(function(promise, track){
+                      if (++nextTrackNo > 1) {
+                        return promise.then(() => player.coordinator.addURIToQueue(track.uri, track.metadata, true, nextTrackNo));
+                      } else {
+                        return promise.then(() => {});
+                      }   
+                    }, Promise.resolve());
```

Comment:
☺ The +1 is to skip the first track that has already been queued.  The intent is to give a quick response by queuing the first track and start it playing immediately while the rest of the queue is loaded.  Avoids you sitting there for a couple minutes wondering if anything is going to happen or not.

BTW, more info on the rogue “member”.  Object keys are $attrs,satellite.  No uuid. Where do I capture the XML that you were hoping to get?


[cid:image001.gif@01D1B743.4FF170D0]<http://www.infor.com/>

Jim Plourde | VP Cloud Services
mobile: 864.270.7677 | office: 864.546.5089 | jim.plourde@infor.com<mailto:jim.plourde@infor.com> | http://www.infor.com<http://www.infor.com/>



From: Jimmy Shimizu <notifications@github.com>
Reply-To: jishi/node-sonos-http-api <reply@reply.github.com>
Date: Thursday, May 26, 2016 at 11:05 AM
To: jishi/node-sonos-http-api <node-sonos-http-api@noreply.github.com>
Cc: James Plourde <Jim.Plourde@infor.com>, Author <author@noreply.github.com>
Subject: Re: [jishi/node-sonos-http-api] Finished and tested Promises appleMusic (#215)


In lib/actions/appleMusic.js<https://github.com/jishi/node-sonos-http-api/pull/215#discussion_r64762416>:

> +                  return tracksArray;

> +                }, []);

> +

> +                nextTrackNo = 0;

> +                return player.coordinator.clearQueue()

> +                  .then(() => player.coordinator.setAVTransport(queueURI, ''))

> +                  .then(() => player.coordinator.addURIToQueue(queueTracks[0].uri, queueTracks[0].metadata, true, 1))

> +                  .then(() => player.coordinator.play())

> +                  .then(() => {

> +                    return queueTracks.reduce(function(promise, track){

> +                      if (++nextTrackNo > 1) {

> +                        return promise.then(() => player.coordinator.addURIToQueue(track.uri, track.metadata, true, nextTrackNo));

> +                      } else {

> +                        return promise.then(() => {});

> +                      }

> +                    }, Promise.resolve());

I would change it myself but won't be able to test if it really works though :)

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<https://github.com/jishi/node-sonos-http-api/pull/215/files/9d1ee91c9a414c74dbf91a4f9d1c9cf4a63b1296#r64762416>